Fix #882 text search feature

PhotoNomad0 opened the pull request:
Fixes #882 text search feature

Changes in this pull request:
- Added basic search filter to chunk and review modes.
- Added drop down list to toggle between source and translation search.

@neutrinog - have a simple search bar here. The search will only be available in chunk and review modes. Didn't add search to read mode since this would take a significant amount of rework to the adapter. Didn't seem necessary since the user can search in chunk modes. Also finding a word on a large chapter seems less useful since the user would still have to look through a lot of text to find the word.

PhotoNomad0 commented:
@neutrinog - looking at this again.  Discovered an UI bug that closing search with the x button is not clearing the search filter.  Looking into it.

PhotoNomad0 commented:
Changes:
- Cards are removed from display while search is running.
- added spinning progressbar for when search is running.

Fixes added: 
- Clear search text on search bar close and clear search filter
- Clear search on switching between modes: read, chunked, review.

Added highlight text in chunk mode, changes for this:
- added flags to ListItem to keep track if card has been rendered with search text highlighting.  This way I know which cards need to be re-rendered when search text changes or is cleared.  This speeds things up greatly when search string has changed by eliminating unnecessary re-rendering.

Tablet limitations:
- can be slow highlighting text in large books like psalms
- search only works in chunk and review modes.

@neutrinog - please take another look at this PR with search highlighting and search spinner added.

neutrinog commented on app/build.gradle:
```diff
@@ -5,7 +5,7 @@ buildscript {
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.1.2'
+        classpath 'com.android.tools.build:gradle:2.1.3'
```
Please revert this back to 2.1.2. We've had trouble with updating gradle to the latest and greatest due to some of the library dependencies we have. E.g. tests or libraries can fail. If I remember correctly we had to stay with gradle 2.1.2 in order for tests to pass on travis.

neutrinog replied:
Ok, I re-ran the tests on travis and they seem to be working. So this probably isn't affecting the tests. However, let's still revert it unless you need 2.1.3 for something.

neutrinog commented on app/src/main/java/com/door43/translationstudio/newui/translate/ChunkModeAdapter.java:
```diff
@@ -993,5 +1010,217 @@ public boolean isFrame() {
+     * @return
+     */
+    private Frame loadFrame(String chapterSlug, String frameSlug) {
+        String complexSlug = chapterSlug + "-" + frameSlug;
```
complexSlug is not used in this context.

neutrinog commented on app/src/main/java/com/door43/translationstudio/newui/translate/ChunkModeAdapter.java:
```diff
@@ -993,5 +1010,217 @@ public boolean isFrame() {
+     * clear the cards displayed since we have new search string
+     * @return
+     */
+    private void clearCardDisplay() {
```
This is only used once and it's only 2 lines of code so this would be better suited inline. If you want to add clarity of purpose you can add a one line comment like
```
# clear card display for search
mFilteredItems = new ListItem[0];
notifyDataSetChanged();
```

neutrinog commented on app/src/main/java/com/door43/translationstudio/newui/translate/ReviewModeAdapter.java:
```diff
@@ -1941,4 +1999,152 @@ public void loadTranslations(SourceTranslation sourceTranslation, TargetTranslat
+     * clear the cards displayed since we have new search string
+     * @return
+     */
+    private void clearCardDisplay() {
```
This is also used only once.

neutrinog commented on app/src/main/java/com/door43/translationstudio/newui/translate/TargetTranslationActivity.java:
```diff
@@ -300,6 +349,240 @@ public boolean onMenuItemClick(MenuItem item) {
+    /**
+     * do search with search string in edit control
+     */
+    public void setSearchFilter() {
```
This is used only once and is just one line. It does not need to be in a function.

neutrinog commented on app/src/main/java/com/door43/translationstudio/newui/translate/TargetTranslationActivity.java:
```diff
@@ -300,6 +349,240 @@ public boolean onMenuItemClick(MenuItem item) {
+    /**
+     * see if translation search is selected
+     */
+    private boolean isTranslationSearchSelected() {
```
This is only used once and does not need to be a function.
We could perform this same action inline like 

```
boolean searchTarget = getSearchTypeSelection() == 1;
```

neutrinog commented on build.gradle:
```diff
@@ -5,7 +5,7 @@ buildscript {
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.1.2'
+        classpath 'com.android.tools.build:gradle:2.1.3'
```
revert to 2.1.2

PhotoNomad0 commented:
@neutrinog  - Thanks for the feedback.  Have made the suggested corrections.

PhotoNomad0 commented:
@neutrinog - this now has the pair programming changes we made.  Also added rendering to titles in Chunk view to enable search highlighting.  Added search highlighting to default renderer.
